### ISSUE TITLE
Proposal: vendor all Go code dependencies in this directory

### ISSUE THREAD

**Issue by arschles** [kubernetes-incubator/service-catalog#59]
Currently, we have glide files, but the code dependencies (in the `vendor` directory) are not checked in. To bring this repository closer in structure and style with kubernetes/kubernetes, I propose checking in the vendor directory alongside the glide files.

cc/ @pmorie @martinmaly @krancour

**Comment by martinmaly** [comment]
other than increasing the size of the repository by several orders of magnitude (the dependency footprint is > 1M loc) I am not sure why we should use two different systems of managing dependencies, in parallel. Either will suffice on their own, no?

if we decide to checkin the dependencies I wonder why glide would be needed, and vice versa.

**Comment by arschles** [comment]
@martinmaly if we do 1 or the other, we will be departing from the kubernetes/kubernetes structure (although that repository uses Godeps).

My preference would be to keep the glide files and nothing else, but that adds another step for new developers to get started (`glide install`). Also, I think @pmorie has additional arguments for at least adding a `vendor` directory.

Disclosure: my bias is heavily toward glide

**Comment by duglin** [comment]
If we build "exclusively" under Docker then there's nothing to install - not even go.  Just docker. That's my preferred path.

**Comment by martinmaly** [comment]
my bias is also toward glide, and no vendor directory.

re: Docker build. the only thing I disagree with is "exclusively". On Linux it will make less of a difference but the cross-compiling builds for Darwin target are significantly slower under Docker than on raw mac.

Additionally, building exclusively inside docker makes it difficult to run the CI pipeline inside containers (nested Docker doesn't really work well). That's some of the complexity in the makefiles where our CI pipeline which runs on k8s cannot run nested Docker build.

Therefore, I think we need to support both.

**Comment by arschles** [comment]
@duglin can you open another issue with details on how the docker-based build system should work?

Like @martinmaly I'd like us to offer the docker-based build as an option, but I don't think it should be exclusive, and I'd like to scope this issue to just talking about how we want to manage our dependencies for host based builds

**Comment by martinmaly** [comment]
Please keep in mind that Docker-based build is already supported (albeit maybe it doesn't meet some requirements).

**Comment by vaikas-google** [comment]
+1 to glide. I really don't want to be checking in that much code to the repo.

**Comment by duglin** [comment]
what's the benefit of glide over just a script that pulls down each package at a certain commit/version?

**Comment by arschles** [comment]
if we use glide we don't have to write that script 😉

**Comment by duglin** [comment]
LOL I was hoping for a little more info since its not just a glide.yml file, it makes a glide.lock so I figured there must be some kind of "smarts" going on there.

**Comment by arschles** [comment]
Ah, got it. Sorry for that terse-ness. Here are a few more specific advantages I can think of:

- dependency graph traversal
- version updates (the `.lock` file represents what the script will go and fetch, the `.yaml` file represents the package whose version can be upgraded)
- CLI tooling for dependency lifecycle management
- easy ability to get packages not inside `vendor` (helpful when running `go test`, for example)

**Comment by duglin** [comment]
ok dependency graph traversal would be nice

**Comment by duglin** [comment]
thanks!

**Comment by MHBauer** [comment]
glide seems to be the hot new thing in golang. I vote we switch to java.

**Comment by duglin** [comment]
Yea, I mean why reinvent the wheel.... Java has maven, ant, javadocs, .... and real inheritance!!!  I mean come on people!

**Comment by pmorie** [comment]
I would rather vendor dependencies than use glide for the following reasons:

1.  Kubernetes does not use glide
2.  Glide does not mitigate the fact that you will eventually need to download dependencies, only delays the eventual download and gives the illusion of a fast git clone
3.  If you need to carry a patch for a dependency you are out of luck with glide, as far as I can tell

**Comment by arschles** [comment]
@pmorie since k8s has the godeps file, I'd like to keep the `glide.yaml` and `glide.lock` files in place in addition to adding the `vendor/`. cool with that?

**Comment by pmorie** [comment]
@arschles sorry, don't follow the logic.  Elaborate?

**Comment by arschles** [comment]
Kubernetes has the [`Godeps.json` file](https://github.com/kubernetes/kubernetes/blob/master/Godeps/Godeps.json) which tracks what is in their `vendor/` directory, and the `glide.{yaml,lock}` serves the same purpose, so my proposal is to keep those files in this directory to match what kubernetes/kubernetes does more closely.

**Comment by pmorie** [comment]
Yeah, I get it now.  Personally, I would rather just use Godep.  I can live with what you suggest as an intermediate step.  I need to write something well-reasoned about this general make-it-like-kubernetes concern and have a group discussion around it.

**Comment by arschles** [comment]
@pmorie when you do write such an issue, we should also record the fact that we should switch to Godep

**Comment by pmorie** [comment]
Not concerned about this at all right now.
